texi2rst.py

```diff
@@ -1187,10 +1187,11 @@ def before(self, w):
         w.write('\n')
 
     def after(self, w):
-        if len(self.element.children) == 1:
-            if isinstance(self.element.children[0], Text):
-                len_ = len(self.element.children[0].data)
-                w.write('\n%s\n\n' % (self.underline * len_))
+        firstchild = self.element.children[0]
+        length = len(self.element.get_all_text())
+        if isinstance(firstchild, Element) and isinstance(firstchild.rst_kind, MatchedInlineMarkup):
+            length += 2 * len(firstchild.rst_kind.tag)
+        w.write('\n%s\n\n' % (self.underline * length))
 
```

Comment:
What if the inline markup is more complicated than just fully enclosing a single text element?

How about making a temporary RstWriter to a StringIO to capture what the .rst output of the children would be (assuming it's all on one line), and then using that to determine the length of the underline they need? - if that makes sense.  Or to extend RstWriter to capture how much was written, so that before/after hooks can determine how much was written for the children, and size the underline accordingly.

Reply:
Done, good idea!